rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,572 @@
+consistent across the nixpkgs is hard to codify with absolute rule. Instead,
+this RFC provides set of guidelines and examples based on scenarios brought up
+during the discussion.
+
+Feature flags that require single external dependency SHOULD be named after
+that dependency. Prefix `lib` SHOULD be removed. For example,
+
+<details><summary>Example</summary>
+
+```
+systemd => with_systemd
+libgif  => with_gif
+curl    => with_curl
+alsa    => with_alsa
+
+# When people say "Xorg" support, they usually mean linking "X11"
+# library which is one of the ways to talk X protocol to X server. "xcb"
+# being another one.
+xorg    => with_x11
+xcb     => with_xcb
+Qt5     => with_qt5
+
+# If upstream feature name contains multiple words (e.g --enable-frame-pointers)
+# then feature parameters SHOULD be snake_cased after upstream.
+--enable-frame-pointers => enable_frame_pointers
+
+# If optional dependency package name has multiple words in it, feature
+# name SHOULD follow the package attribute, but with dashes replaced
+# with underscores.
+pulseaudio => with_pulseaudio
+dump-init => with_dump_init
+
+
+```
+
+</details>
+
+When multiple feature flags require the same build dependency, for example
+derivation has optional support for FTP and HTTP protocols, any of which
+incur dependency on `curl`, the package would look like following:
+
+<details><summary>Example</summary>
+
+```
+{ lib
+, stdenv
+, curl
+, with_curl ? true
+, enable_ftp ? true
+, enable_http ? true
+}:
+
+# Assertions are fully optional.
+assert enable_ftp -> with_curl;
+assert enable_http -> with_curl;
+
+stdenv.mkDerivation {
+   # ... snip ...
+
+   configureFlags = lib.optionals enable_http ["--enable-http"]
+                 ++ lib.optionals enable_ftp ["--enable-ftp"];
+   buildInputs = lib.optionals withCurl [ curl ];
+
+   # ... snip ...
+
+   passthru.features = {
+      inherit with_curl enable_ftp enable_http;
+   };
+
+   # ... snip ...
+}
+```
+</details>
+
+When build dependency comes from particular package set, it MAY make sense to
+name feature parameter after it. E.g build dependency on `qt6.qtbase` SHOULD
+have `with_qt6` feature parameter. If that results in feature parameter name
+that is too generic, package name from the set MAY be included into the feature
+parameter name. Optional dependency on `pythonPackages.pillow` MAY have feature
+parameter `with_python_pillow`.
+
+Build dependency on bindings to low-level libraries SHOULD be named after
+underlying library. For example, optional dependecy on `pyqt5` Python bindings
+to `Qt5` library should have `with_qt5` feature parameter.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This is how one can query list of feature parameters that incur extra runtime dependencies:
+
+```
+$ nix eval --json --impure --expr 'with import ./. {}; builtins.attrNames gnuplot.override.__functionArgs' | jq '.[]| select(startswith("with"))'
```

Comment:
I would like if we can do it without jq or similar shell script tricks.

But I believe it would require a too huge modification to the language. Nix does not look like a query language.

Reply:
It's not that people have not implemented typed (and checked) arguments in Nix, they have, multiple times. So the massive change would be on Nixpkgs side… and the speed impact will be unpleasant